src/python/pants/backend/python/rules/coverage.py

```diff
@@ -0,0 +1,279 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import configparser
+import json
+from dataclasses import dataclass
+from io import StringIO
+from typing import Dict, Tuple
+import itertools
+
+import pkg_resources
+
+from pants.backend.python.rules.inject_init import InjectedInitDigest
+from pants.backend.python.rules.pex import (
+  CreatePex,
+  Pex,
+  PexInterpreterConstraints,
+  PexRequirements,
+)
+from pants.backend.python.rules.prepare_chrooted_python_sources import ChrootedPythonSources
+
+from pants.backend.python.rules.python_test_runner import (
+  DEFAULT_COVERAGE_CONFIG,
+  get_coveragerc_input,
+)
+from pants.backend.python.subsystems.pytest import PyTest
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.backend.python.subsystems.python_tool_base import PythonToolBase
+from pants.backend.python.subsystems.subprocess_environment import SubprocessEncodingEnvironment
+from pants.base.specs import AddressSpecs
+from pants.build_graph.address import Address
+from pants.engine.addressable import BuildFileAddresses
+from pants.engine.fs import (
+  Digest,
+  DirectoriesToMerge,
+  DirectoryWithPrefixToAdd,
+  FileContent,
+  FilesContent,
+  InputFilesContent,
+)
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.legacy.graph import HydratedTarget, TransitiveHydratedTargets, HydratedTargets
+from pants.engine.rules import goal_rule, rule, subsystem_rule
+from pants.engine.selectors import Get, MultiGet
+from pants.rules.core.strip_source_root import SourceRootStrippedSources
+from pants.rules.core.test import AddressAndTestResults, PytestCoverageReport
+from pants.source.source_root import SourceRootConfig
+
+
+class CoverageToolBase(PythonToolBase):
+  options_scope = 'pytest-coverage'
+  default_version = 'coverage==5.0.0'
```

Comment:
It can be anything >= 5, right? If so, we've recently started floating these requirement strings so that Pants no longer has to worry about staying up to date with new releases in libraries like Pytest and Flake8. So, this would be `coverage>=5.0.0`.

Reply:
So, the coverage library has a history of breaking changes between minor versions (eg: changing the data format from json to sql) It's *super* important that `python_test_runner`, `merge_coverage_data` and `create_coverage_report` all use exactly the same version of coverage, since they're all touching the same data and that data must be in the same format for each of them. 

I think it is moderately unlikely we'll have another breaking change any time soon, but I did run into issues with pex resolving different versions of coverage for each of the rules being executed hence pinning it. `coverage~=5.0.0` is probably safe.

Reply:
Agreed on pinning. We should have lockfiles. But in the absence of lockfiles, floating versions are always risky.